tools/buildtools/schc.py

```diff
@@ -0,0 +1,923 @@
+    TokenType.IF_SINCETIME:         TokenProperties(isConditionalBranch=True,   hasArguments=True, isGeneric=True,                     shortVersion=TokenType.IF_SINCETIME_S, longVersion=TokenType.IF_SINCETIME_L),
+    TokenType.BRANCH:               TokenProperties(isUnconditionalBranch=True, hasArguments=True, isGeneric=True,                     shortVersion=TokenType.BRANCH_S, longVersion=TokenType.BRANCH_L),
+
+    # Extra tokens
+    TokenType.ELSE:                 TokenProperties(isExtraToken=True),
+    TokenType.BRACE_OPEN:           TokenProperties(isExtraToken=True),
+    TokenType.BRACE_CLOSE:          TokenProperties(isExtraToken=True),
+    TokenType.ARGS:                 TokenProperties(isExtraToken=True),
+    TokenType.NOT:                  TokenProperties(isExtraToken=True),
+    TokenType.LABEL:                TokenProperties(isExtraToken=True),
+}
+
+tokenLiterals: dict[str, TokenType] = {x.value: x for x in TokenType}
+
+
+regex_label = re.compile(r"(?P<label>\w+)\s*:")
+regex_identifier = re.compile(r"(?P<identifier>\w+)")
+regex_individualTokens = re.compile(r"(?P<individual>[\{\}])")
+
+@dataclasses.dataclass
+class Token:
+    tokenType: TokenType
+
+    # The literal read from the input. Some token types don't have a fixed literal, like LABEL or ARGS
+    tokenLiteral: str
+
+    # Track the token position on the original file, for better error messages
+    filename: str
+    lineNumber: int
+    columnNumber: int
+
+    def getProperties(self) -> TokenProperties:
+        return tokenPropertiesDict[self.tokenType]
+
+    def newFromTokenType(self, newType: TokenType) -> Token:
+        return Token(newType, newType.value, self.filename, self.lineNumber, self.columnNumber)
+
+    def newFromTokenTypePreserveLiteral(self, newType: TokenType) -> Token:
+        return Token(newType, self.tokenLiteral, self.filename, self.lineNumber, self.columnNumber)
+
+    def __str__(self) -> str:
+        ret = f"{self.filename}:{self.lineNumber}:{self.columnNumber}\n"
+        ret += f"  {self.tokenType.name}"
+        if self.tokenLiteral != self.tokenType.value:
+            ret += f" '{self.tokenLiteral}'"
+        return ret
+
+
+# Allows to know which token will be the next one, even with recursive functions
+class TokenIterator:
+    def __init__(self, tokens: list[Token]):
+        self.tokens = list(tokens)
+        self.index = 0
+
+    def get(self) -> Token|None:
+        if self.index >= len(self.tokens):
+            return None
+        token = self.tokens[self.index]
+        self.index += 1
+        return token
+
+    def unget(self) -> None:
+        self.index -= 1
+        if self.index < 0:
+            raise RuntimeError()
+
+    def remainingTokens(self) -> int:
+        return len(self.tokens) - self.index
+
+    def reset(self) -> None:
+        self.index = 0
+
+
+# Strips comments
+def preprocess(contents: str, filename: str) -> str:
+    result = ""
```

Comment:
ftr (not an issue here because the schedule scripts will never be big) appending characters one at a time to a string in python is quadratic complexity, i.e. if `s` is of length N>>1 then
```py
r = ""
for c in s:
  r += c
```
takes a time proportional to N^2

Because the string is reallocated and copied every loop. CPython can sometimes optimize this but YMMV

What you want to do in python is build a list of characters and concatenate all at once, it's typically unexpectedly faster
```py
L = []
for c in s:
  L.append(c)
r = "".join(L)
```

Again in this case it's fine, you're processing less than, idk, ~1000 characters per string ever

Reply:
Compiling all the scripts with no changes to the compiler
```
real    0m4.517s
user    0m3.518s
sys     0m0.852s
```
With this suggestion:
```
real    0m4.437s
user    0m3.436s
sys     0m0.927s
```

Since the files are so small seems to not really matter much. But thanks for pointing it out, this can help in other tools of mine like spimdisasm lol